pkg/provisioners/gcppubsub/dispatcher/receiver/cache/cache.go

```diff
@@ -0,0 +1,216 @@
+	if present {
+		current.extend(c.currentTime, c.ttl)
+		return current.value
+	} else {
+		c.items[k] = newItem
+		return value
+	}
+}
+
+// Get gets an item in the cache and extends its time to live. If the key is not present in the
+// cache, nil is returned.
+func (c *TTL) Get(keyString string) Stoppable {
+	k := key(keyString)
+	c.itemsLock.Lock()
+	defer c.itemsLock.Unlock()
+	i := c.items[k]
+	if i != nil {
+		i.extend(c.currentTime, c.ttl)
+		return i.value
+	}
+	return nil
+}
+
+// extend extends the time this item's time to live to now plus ttl.
+func (i *item) extend(currentTime func() time.Time, ttl time.Duration) {
+	i.ttl = currentTime().Add(ttl)
+}
+
+// DeleteIfPresent deletes an entry from the cache if its value in the cache is equal to value.
+func (c *TTL) DeleteIfPresent(keyString string, value Stoppable) {
+	stoppable := func() Stoppable {
+		c.itemsLock.Lock()
+		defer c.itemsLock.Unlock()
+		k := key(keyString)
+		actualValue, present := c.items[k]
+		if present {
+			if actualValue.value == value {
+				delete(c.items, k)
+				return value
+			}
+		}
+		return nil
+	}()
+	if stoppable != nil {
+		stoppable.Stop()
+	}
+}
+
+// Start continuously culls the old items from the cache. It should be run in a goroutine.
+func (c *TTL) Start(stopCh <-chan struct{}) error {
+	for {
+		select {
+		case <-stopCh:
+			c.deleteAllItems()
+			return nil
+		case <-time.After(c.reapingPeriod):
+			c.cull()
+		}
+	}
+}
+
+// cull all items from the cache that are past their time to live.
+func (c *TTL) cull() {
+	// Minimize time when the lock is held by collecting all the functions that will be stopped, but
+	// not running them, under lock.
+	stops := func() []func() {
+		stops := make([]func(), 0, stopSliceSize)
+		now := c.currentTime()
+		c.itemsLock.Lock()
+		defer c.itemsLock.Unlock()
+		for k, i := range c.items {
+			if i.ttl.Before(now) {
+				stops = append(stops, c.deleteUnderLock(k))
+			}
+		}
+		return stops
+	}()
+	for _, stop := range stops {
+		stop()
+	}
+}
+
+// deleteUnderLock deletes the item from the cache, if it is present. It returns the Stop() function
+// that should be called to actually stop the entry.
+//
+// This function can only be called if the c.itemsLock lock is held.
+func (c *TTL) deleteUnderLock(k key) func() {
+	i, present := c.items[k]
+	if present {
+		delete(c.items, k)
+		return i.value.Stop
+	}
+	// I don't think this can happen, as this can only be called under lock and should only be
+	// called if the key is present, but just in case return a nop function.
+	return func() {}
+}
+
+// deleteAllItems is expected to be called only when the cache is stopped. It removes all items from
+// the cache and Stop()s them.
+func (c *TTL) deleteAllItems() {
+	oldItems := func() map[key]*item {
+		c.itemsLock.Lock()
+		defer c.itemsLock.Unlock()
+		oldItems := c.items
+		// Any subsequent cache gets should not return any of these items, so just make a new map.
+		c.items = make(map[key]*item)
```

Comment:
why not make it into a nil?

Reply:
I want subsequent `Get`s to return `nil`, not panic. If I set this to `nil`, then `Get` would panic.